Library/include/CSP/Multiplayer/EventBus.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2023 Magnopus LLC
+
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "CSP/Common/String.h"
+#include "CSP/Multiplayer/MultiPlayerConnection.h"
+#include "Common/UUIDGenerator.h"
+#include "Multiplayer/EventSerialisation.h"
+
+#include <map>
+
+namespace csp::systems
+{
+
+class SpaceSystem;
+class SystemsManager;
+class SystemBase;
+
+} // namespace csp::systems
+
+namespace csp::memory
+{
+
+CSP_START_IGNORE
+template <typename T> void Delete(T* Ptr);
+CSP_END_IGNORE
+
+} // namespace csp::memory
+
+
+/// @brief Namespace that encompasses everything in the multiplayer system
+namespace csp::multiplayer
+{
+
+class ReplicatedValue;
+
+/// @ingroup Multiplayer
+/// @brief Handling of all network events.
+class CSP_API EventBus
+{
+public:
+	/** @cond DO_NOT_DOCUMENT */
+	friend class csp::systems::SpaceSystem;
+	friend class csp::systems::SystemsManager;
+	friend class MultiplayerConnection;
+	friend void csp::memory::Delete<EventBus>(EventBus* Ptr);
+	/** @endcond */
+
+	// The callback used to register to listen to events.
+	typedef std::function<void(bool, const csp::common::Array<ReplicatedValue>&)> ParameterisedCallbackHandler;
+
+	/// @brief Registers a system to listen for the named event, where the system can define its
+	/// @brief own callback and deserialiser.
+	/// @param EventName csp::common::String : The identifying name for the event to listen for.
+	/// @param System csp::systems::SystemBase* : A pointer to the system which wants to register for the event.
+	CSP_NO_EXPORT void ListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System);
```

Comment:
could you please confirm whether the use of CSP_NO_EXPORT for the 2 system-related ListenEvent and StopListenEvent methods is correct? @MAG-AdamThorn @MAG-mv

Reply:
confirmed during a call